### PR TITLE
Instanced graphics

### PR DESCRIPTION
**Pull request by pmconne**
Add support for creating reusable graphic templates that can be instanced multiple times.

## Motivation

1. WFS server serving data sets consisting of many, many points indicating the locations of various objects like different kinds of road signs. They have a glTF model for each kind of road sign. They want to efficiently draw each road sign at the position specified in the WFS data (i.e., instanced rendering). They need to be able to process the WFS data and produce the graphics on a Worker. They want to reuse the same glTF assets with different sets of instances without duplicating the WebGL resources.

1. Use `GraphicBuilder` or `GraphicDescriptionBuilder` to create a graphic, which may internally contain several graphics. Draw that graphic repeatedly using instancing. When the graphic contains multiple graphics, the WebGL resources defining the instances should be reused for each graphic - not duplicated.

## Obstacles

Today we have `InstancedGraphicParams` which can be used to specify that you want to draw the same geometry repeatedly with different transforms, feature Ids, and symbology. We use it for iMdl, glTF, and particle collections. However, we provide no APIs to create an `InstancedGraphicParams` directly, nor documentation of how to do it yourself. InstancedGraphicParams are not reusable - they allocate WebGL buffers that are associated with one graphic.

## Solution

Define the concept of a `GraphicTemplate`. You can obtain one from a GraphicBuilder, GraphicDescriptionBuilder, or glTF asset. In its simplest form it is just a list of `RenderGeometry`. Since glTF assets describe a scene graph, the templates they produce may contain multiple nodes, each consisting of a list of RenderGeometry and a Transform.

Define the concept of `RenderInstances`. This defines the transforms and optional feature Ids and/or symbology overrides for a set of instances.

Add `RenderInstancesParams` describing how to create a `RenderInstances`, and a straightforward, Worker-safe `RenderInstancesParamsBuilder` API to produce them.

Add to `RenderSystem` the following operations:
- `createGraphicTemplate`. This is used internally by the public APIs.
- `createRenderInstances` to produce a `RenderInstances` (holding reusable WebGL buffers) from a `RenderInstancesParams`
- `createGraphicFromTemplate` accepting a GraphicTemplate and a RenderInstances and producing a RenderGraphic.

This decouples the instances from the template, so they can be mixed, matched, and reused as the user sees fit.

I enhanced display-test-app's `dta gltf` key-in with additional arguments to support instancing.

## TODO

- [x] NextVersion.md
- [x] Example code
- [x] ImageTests - no differences reported
- [x] Unit tests
- [x] Update display-test-app README
- [x] Fix vite run-time error in display-test-app

### CODE REVIEW

File: core/frontend/src/render/webgl/InstancedGeometry.ts [hunk 1]
```diff
@@ -69,23 +71,105 @@ export interface PatternTransforms {
...
+    if (undefined !== symbologyOverrides && undefined === (symBuf = BufferHandle.createArrayBuffer(symbologyOverrides)))
+      return undefined;
+
+    const tfBuf = BufferHandle.createArrayBuffer(transforms);
+    const transformCenter = params.transformCenter instanceof Point3d ? params.transformCenter : Point3d.fromJSON(params.transformCenter);
+    if (!tfBuf) {
+      return undefined;
+    }
+
+    return undefined !== tfBuf ? new InstanceBuffersData(count, tfBuf, transformCenter, symBuf, idBuf, disableDisposal) : undefined;
```
Comment (DStradley) [inline core/frontend/src/render/webgl/InstancedGeometry.ts:110]: Aren't you already checking `tfBuf` right above this?

**Review by DStradley — APPROVED** [review]

File: core/frontend/src/render/webgl/InstancedGeometry.ts (hunk 1 above)
Comment (pmconne) [reply core/frontend/src/render/webgl/InstancedGeometry.ts:110]: Fixed.

File: core/frontend/src/render/RenderSystem.ts [hunk 2]
```diff
@@ -787,10 +855,17 @@ export abstract class RenderSystem implements IDisposable {
   /** Convert a [[GraphicDescription]] produced by a [[GraphicDescriptionBuilder]] into a [[RenderGraphic]].
    * @beta
    */
-  public async createGraphicFromDescription(args: CreateGraphicFromDescriptionArgs): Promise<RenderGraphic | undefined> {
+  public createGraphicFromDescription(args: CreateGraphicFromDescriptionArgs): RenderGraphic  | undefined {
```
[the remark this answers is not in the document]
Comment (pmconne) [reply core/frontend/src/render/RenderSystem.ts]: Thanks. Fixed. FYI you can use the "add suggestion" button to make these trivial changes easy to address.
image

**Review by markschlosseratbentley — APPROVED** [review]

**Review by ben-polinsky — APPROVED** [review]

File: core/frontend/src/render/webgl/InstancedGeometry.ts (hunk 1 above)
Comment (pmconne) [reply core/frontend/src/render/webgl/InstancedGeometry.ts]: Please resolve comments when you are satisfied.

**Comment by pmconne** [comment]
display-test-app builds successfully (`rush build --to display-test-app`) but at run-time produces the following error in browser console:

Uncaught SyntaxError: The requested module '/@fs/home/paul/c/js/s/itwinjs/core/frontend/src/common/render/InstancedGraphicParams.ts' does not provide an export named 'InstancedGraphicParams' (at ImdlGraphicsCreator.ts:23:10)

Seems like a vite issue. I have run into this before, and solved it by replacing `import` with `import type`.
However, ImdlGraphicsCreator.ts definitely uses InstancedGraphicParams as both a value and a type, so it needs to import it as a value. InstancedGraphicsParams is both an interface and a namespace.

@aruniverse do you know anybody who knows anything about vite?

**Comment by aruniverse** [comment]
> @aruniverse do you know anybody who knows anything about vite?

@hl662 or I can take a look at this tomorrow

**Comment by pmconne** [comment]
> @hl662 or I can take a look at this tomorrow

Thank you. FWIW, the following appeases it for that file but just produces the same error for the import in RenderInstancesParamsImpl.ts.

```
diff --git a/core/frontend/src/tile/ImdlGraphicsCreator.ts b/core/frontend/src/tile/ImdlGraphicsCreator.ts
index 1cf01eb716..48ea3eed1f 100644
--- a/core/frontend/src/tile/ImdlGraphicsCreator.ts
+++ b/core/frontend/src/tile/ImdlGraphicsCreator.ts
@@ -20,7 +20,7 @@ import { VertexIndices } from "../common/internal/render/VertexIndices";
 import type { RenderGraphic } from "../render/RenderGraphic";
 import { GraphicBranch } from "../render/GraphicBranch";
 import type { RenderSystem } from "../render/RenderSystem";
-import { InstancedGraphicParams } from "../common/render/InstancedGraphicParams";
+import * as instancedGraphicParams from "../common/render/InstancedGraphicParams";
 import type { IModelConnection } from "../IModelConnection";
 import { GraphicDescription } from "../common/render/GraphicDescriptionBuilder";
 import { GraphicDescriptionImpl, isGraphicDescription } from "../common/internal/render/GraphicDescriptionBuilderImpl";
@@ -226,11 +226,11 @@ function getMaterial(mat: string | Imdl.SurfaceMaterialParams, options: Graphics
   return options.system.createMaterial(params, options.iModel);
 }

-function getModifiers(primitive: Imdl.Primitive): { viOrigin?: Point3d, instances?: InstancedGraphicParams } {
+function getModifiers(primitive: Imdl.Primitive): { viOrigin?: Point3d, instances?: instancedGraphicParams.InstancedGraphicParams } {
   const mod = primitive.modifier;
   switch (mod?.type) {
     case "instances":
-      return { instances: InstancedGraphicParams.fromProps(mod) };
+      return { instances: instancedGraphicParams.InstancedGraphicParams.fromProps(mod) };
```

**Comment by pmconne** [comment]
The error occurs in electron. It does not occur (and the app runs normally) in firefox.

**Comment by pmconne** [comment]
> Are there any other steps to reproduce?

Not as far as I can tell.
```
rush build --to display-test-app
cd test-apps/display-test-app
npm run start
```

Opens the window with infinite spinner and error in dev tools console.
I'm not sure what variables between our setups might be relevant. Ubuntu here, but I'd hope that wouldn't matter. I can test on Windows though.

I know little about vite. When I run the dev server is it compiling on the fly? Is it putting the compiled code somewhere I can inspect it?

**Comment by aruniverse** [comment]
> I know little about vite. When I run the dev server is it compiling on the fly? Is it putting the compiled code somewhere I can inspect it?

When running the dev server, Vite is compiling the app on the fly with esbuild. We added a vite plugin, [vite-plugin-inspect](https://www.npmjs.com/package/vite-plugin-inspect), to debug but it might not be the most useful. You can see all the transforms Vite runs (actually make the code ESM compliant, eg changes imports to use the actual file extension). While the dev server is running you can visit `http://localhost:3000/__inspect`, you can see in memory transformed files 

image

http://localhost:3000/__inspect/#/module?id=%2FUsers%2Farun%2Frepos%2Fgh%2Fimjs%2Fitwinjs-core%2Fcore%2Ffrontend%2Fsrc%2Fcommon%2Frender%2FInstancedGraphicParams.ts
image

Will investigate if theres anything more we can do to help debug

**Comment by pmconne** [comment]
Yeah, inspect shows the same for me. It's clearly exporting `InstancedGraphicParams` and lets me access it if I `import * as Foo`.

**Comment by pmconne** [comment]
I can't reproduce the issue on Windows either. The contents of InstancedGraphicParams.ts and ImdlGraphicsCreator.ts look identical to me in vite-inspect, other than the absolute paths differing as expected.

File: core/frontend/src/common/render/InstancedGraphicParams.ts [hunk 3]
```diff
@@ -46,6 +44,75 @@ export interface InstancedGraphicParams {
   readonly range?: Range3d;
 }
 
+/** A representation of an [[InstancedGraphicParams]] that can be copied using [structured cloning](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone),
+ * e.g., when transferring between Workers. This representation is used by [[RenderInstancesParams]].
+ * @see [[InstancedGraphicParams.toProps]] and [[InstancedGraphicParams.fromProps]] to convert between representations.
+ * @see [[InstancedGraphicProps.collectTransferables]] to extract Tranferable objects for structured clone.
```
Comment (aruniverse) [inline core/frontend/src/common/render/InstancedGraphicParams.ts:50]:
```suggestion
 * @see [[InstancedGraphicProps.collectTransferables]] to extract Transferable objects for structured clone.
```

**Comment by aruniverse** [comment]
> I can't reproduce the issue on Windows either. 

I'm on Mac. I dont really understand why this would come down to being an OS issue.
Could you provide more details re your dev env? OS ver, node ver, etc

**Comment by pmconne** [comment]
> I'm on Mac. I dont really understand why this would come down to being an OS issue. Could you provide more details re your dev env? OS ver, node ver, etc

Ubuntu 22.04.4 LTS. node v18.16.0. npm 9.5.1.

The same issue is reproducible when importing InstancedGraphicProps from same file. It is a `type` merged with a `namspace`, just as InstancedGraphicParams is an `interface` and `namespace`. I don't see any other examples of such merged type+namespace in core-frontend. Plenty of merged class+namespace, which present no problem, presumably because both are values. I can import+use `ElementGeometryChange` from core-common without complaint - it's a type+namespace.

**Comment by hl662** [comment]
Just following up, I'm on Mac, I can reproduce the error you're getting @pmconne. I'm using node 20.11.1, npm v 10.2.4. Just started investigating now...

**Comment by pmconne** [comment]
I have a second copy of the repo on my linux box, I can repro it in there too. (I thought maaaybe some kind of weird caching issue but it seems not).

**Comment by hl662** [comment]
> I have a second copy of the repo on my linux box, I can repro it in there too. (I thought maaaybe some kind of weird caching issue but it seems not).

Huh. I just updated vite in DTA's package.json to "^5.4.6" and the error is gone. Can you do the same?

**Comment by pmconne** [comment]
> I just updated vite in DTA's package.json to "^5.4.6" and the error is gone. Can you do the same?

Waddayaknow, guess they fixed it. Thanks!

**Pull request merged**